independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/LockInterceptor.java

```diff
@@ -0,0 +1,90 @@
+package io.quarkus.arc.impl;
+
+import static javax.interceptor.Interceptor.Priority.PLATFORM_BEFORE;
+
+import io.quarkus.arc.ArcInvocationContext;
+import io.quarkus.arc.Lock;
+import io.quarkus.arc.LockException;
+import java.lang.annotation.Annotation;
+import java.util.Set;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Priority;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+@Lock
+@Interceptor
+@Priority(PLATFORM_BEFORE)
+public class LockInterceptor {
+
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+
+    @AroundInvoke
+    Object lock(InvocationContext ctx) throws Exception {
+        Lock lock = getLock(ctx);
+        switch (lock.value()) {
+            case WRITE:
+                return writeLock(lock, ctx);
+            case READ:
+                return readLock(lock, ctx);
+            case NONE:
+                return ctx.proceed();
+        }
+        throw new LockException("Unsupported @Lock type found on business method " + ctx.getMethod());
+    }
+
+    private Object writeLock(Lock lock, InvocationContext ctx) throws Exception {
+        boolean locked = true;
+        long time = lock.time();
+        try {
+            if (time > 0) {
+                locked = readWriteLock.writeLock().tryLock(time, lock.unit());
```

Comment:
If this throws InterruptedException then locked will be true but the lock will not have been acquired, and the finally block will attempt to unlock a lock it does not hold.

Reply:
I see. In that case `locked` should be initialized with `false` and set to `true` after the lock is acquired sucessfully.

Reply:
Should be fixed now..